warp-core-macros/src/main/scala/com/workday/warp/monadic/WarpMacros.scala

```diff
@@ -0,0 +1,149 @@
+    // splice (in the context of reify) converts from scala AST -> value, ie Expr[Int] => Int
+    reify {
+      a.splice + b.splice
+    }
+  }
+
+
+  /**
+    * Rewrites a WarpScript for-comprehension AST to automatically insert TestIds into the `measure` calls.
+    *
+    * Constructs a Test Id as the concatenation of the fully qualified enclosing class name with the name of the user-provided
+    * variable for that stage of the [[WarpScript]].
+    *
+    * For example, the following code snippet (assuming it is defined within a class named `com.workday.warp.Test`):
+    *
+    * {{{
+    *   for {
+    *     a <- exec(1 + 1)
+    *     b <- measure(a + 1)
+    *     c <- measure(b + 1)
+    *     d <- measure(c + 1)
+    *    } yield d
+    * }}}
+    *
+    * would be transformed to this snippet:
+    *
+    * {{{
+    *   for {
+    *     a <- exec(1 + 1)
+    *     b <- measure("com.workday.warp.Test.b", a + 1)
+    *     c <- measure("com.workday.warp.Test.c", b + 1)
+    *     d <- measure("com.workday.warp.Test.d", c + 1)
+    *    } yield d
+    * }}}
+    *
+    * This macro should be considered experimental.
+    *
+    * @param script [[WarpScript]] that will be rewritten.
+    * @tparam T underlying return type of `script`.
+    * @return
+    */
+  def deriveTestIds[T](script: WarpScript[T]): WarpScript[T] = macro deriveTestIdsImpl[T]
+
+  def deriveTestIdsImpl[T: ctx.WeakTypeTag](ctx: blackbox.Context)(script: ctx.Expr[WarpScript[T]]): ctx.Expr[WarpScript[T]] = {
+    // this untypecheck call is crucial to avoid crashing the compiler with mysterious errors like
+    // [Error] : Error while emitting MacrosSpec.scala
+    // value c
+    // https://github.com/scala/bug/issues/11628 is possibly related?
+    ctx.Expr[WarpScript[T]](testIdTransformer(ctx).transform(ctx.untypecheck(script.tree)))
+  }
+
+
+  /**
+    * Creates a Tree transformer that traverses a [[WarpScript]] AST and inserts test Id into `measure` calls.
+    *
+    * @param ctx blackbox macro Context.
+    * @return a (path-dependent) Tree transformer.
+    */
+  def testIdTransformer(ctx: blackbox.Context): ctx.universe.Transformer = {
+    import ctx.universe._
+
+    new Transformer {
+      val enclosingClass: String = extractEnclosingClass(ctx)
+      // TODO try to avoid using a var here
+      var capturedArg: Option[Tree] = None
+
+      override def transform(tree: Tree): Tree = tree match {
+        // capture arg here, but we dont need to transform
+        case Apply(func, args) if show(func).startsWith("com.workday.warp.monadic.WarpAlgebra.measure")
+            && (func.symbol.toString == "method map" || func.symbol.toString == "method flatMap") =>
+          capturedArg = args.headOption
+          super.transform(tree)
+
+        // a `measure` call. use the identifier we previously parsed to expand into a TestId
+        case Apply(func, args) if show(func).startsWith("com.workday.warp.monadic.WarpAlgebra.measure")
```

Comment:
Does it make sense to parameterize these string literals (or otherwise turn them into a constant)? Perfectly fine if not.

Reply:
cool i can do that